File: core/gurbanilens/corpus.py
"""SGGS corpus loader.

Wraps the Shabad OS SQLite (v4 schema) and exposes the subset of data Phase 1
needs: SGGS lines with Ang, Pangti, line type, raw Gurmukhi (Anmol Lipi font
encoding), and English transliteration.

Phase 1 design notes:
- `gurmukhi` is stored in Anmol Lipi font encoding (ASCII designed to render as
  Gurmukhi via the Anmol Lipi typeface). Proper Unicode conversion is deferred —
  to be done via anvaad-js or equivalent established library, not hand-rolled.
- `transliteration_en` (from the `transliterations` table) is hand-curated Latin
  romanization. This is the primary matching surface for Phase 1 — both the
  corpus and ASR output get normalized to Latin space before fuzzy matching.
- `pronunciation` on `lines` is empty for SGGS in this DB, so we ignore it.

The matcher pulls all_lines() once to build an in-memory index. The CLI uses
lookup() to fetch a specific line for display.
"""

from __future__ import annotations

import sqlite3
from dataclasses import dataclass
from pathlib import Path
from typing import Iterator

DEFAULT_DB_PATH = Path(__file__).resolve().parent.parent.parent / "data" / "sggs" / "database.sqlite"
SGGS_NAME_PREFIX = "Sri Guru Granth Sahib"
TRANSLITERATION_LANGUAGE = "English"


@dataclass(frozen=True, slots=True)
class Line:
    """A single line of SGGS — Pankti, Rahao, Sirlekh, or Manglacharan."""

    id: str
    shabad_id: str
    ang: int
    pangti: int | None
    line_type: str | None
    gurmukhi: str
    transliteration_en: str | None
    first_letters: str | None
    order_id: int

    def __str__(self) -> str:
        loc = f"Ang {self.ang}"
        if self.pangti is not None:
            loc += f", Pangti {self.pangti}"
        if self.line_type:
            loc += f" [{self.line_type}]"
        body = self.transliteration_en or self.gurmukhi
        return f"{loc}\n  {body}"


_LINE_COLS = """
    l.id, l.shabad_id, l.source_page, l.source_line, l.gurmukhi,
    l.first_letters, l.order_id,
    lt.name_english AS line_type,
    t.transliteration AS transliteration_en
"""


class Corpus:
    """Read-only SGGS corpus backed by Shabad OS SQLite."""

    def __init__(self, db_path: Path | str = DEFAULT_DB_PATH) -> None:
        self.db_path = Path(db_path)
        if not self.db_path.exists():
            raise FileNotFoundError(
                f"Corpus DB not found at {self.db_path}. "
                f"Run: python scripts/fetch_corpus.py"
            )
        self._conn = sqlite3.connect(f"file:{self.db_path}?mode=ro", uri=True)
        self._conn.row_factory = sqlite3.Row
        self._sggs_source_id = self._resolve_id(
            "sources", "name_english LIKE ? || '%'", SGGS_NAME_PREFIX, "SGGS"
        )
        self._english_language_id = self._resolve_id(
            "languages", "name_english = ?", TRANSLITERATION_LANGUAGE, "English language"
        )

    def _resolve_id(self, table: str, where: str, value: str, label: str) -> int:
        row = self._conn.execute(
            f"SELECT id FROM {table} WHERE {where}",
            (value,),
        ).fetchone()
        if row is None:
            raise RuntimeError(
                f"Could not resolve {label} id from {table} (where {where}, value {value!r}). "
                f"DB may be wrong schema version."
            )
        return int(row["id"])

    def _row_to_line(self, row: sqlite3.Row) -> Line:
        return Line(
            id=row["id"],
            shabad_id=row["shabad_id"],
            ang=row["source_page"],
            pangti=row["source_line"],
            line_type=row["line_type"],
            gurmukhi=row["gurmukhi"],
            transliteration_en=row["transliteration_en"],
            first_letters=row["first_letters"],
            order_id=row["order_id"],
        )

    def __len__(self) -> int:
        return self._conn.execute(
            """
            SELECT COUNT(*) AS n
            FROM lines l
            JOIN shabads s ON l.shabad_id = s.id
            WHERE s.source_id = ?
            """,
            (self._sggs_source_id,),
        ).fetchone()["n"]

    def all_lines(self) -> Iterator[Line]:
        """Yield every SGGS line in canonical order."""
        cur = self._conn.execute(
            f"""
            SELECT {_LINE_COLS}
            FROM lines l
            JOIN shabads s ON l.shabad_id = s.id
            LEFT JOIN line_types lt ON l.type_id = lt.id
            LEFT JOIN transliterations t
                ON t.line_id = l.id AND t.language_id = ?
            WHERE s.source_id = ?
            ORDER BY l.order_id
            """,
            (self._english_language_id, self._sggs_source_id),
        )
        for row in cur:
            yield self._row_to_line(row)

    def lookup(self, ang: int, pangti: int) -> list[Line]:
        """Return all SGGS lines at (ang, pangti). Usually one, occasionally more."""
        cur = self._conn.execute(
            f"""
            SELECT {_LINE_COLS}
            FROM lines l
            JOIN shabads s ON l.shabad_id = s.id
            LEFT JOIN line_types lt ON l.type_id = lt.id
            LEFT JOIN transliterations t
                ON t.line_id = l.id AND t.language_id = ?
            WHERE s.source_id = ? AND l.source_page = ? AND l.source_line = ?
            ORDER BY l.order_id
            """,
            (self._english_language_id, self._sggs_source_id, ang, pangti),
        )
        return [self._row_to_line(row) for row in cur]

    def shabad_lines(self, shabad_id: str) -> list[Line]:
        """All lines belonging to a single Shabad, in order."""
        cur = self._conn.execute(
            f"""
            SELECT {_LINE_COLS}
            FROM lines l
            LEFT JOIN line_types lt ON l.type_id = lt.id
            LEFT JOIN transliterations t
                ON t.line_id = l.id AND t.language_id = ?
            WHERE l.shabad_id = ?
            ORDER BY l.order_id
            """,
            (self._english_language_id, shabad_id),
        )
        return [self._row_to_line(row) for row in cur]

    def close(self) -> None:
        self._conn.close()

    def __enter__(self) -> "Corpus":
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

File: ios/GurbaniLensCore/Sources/GurbaniLensCore/Corpus.swift
import Foundation
import SQLite3

/// Read-only SGGS corpus backed by the Shabad OS SQLite (v4 schema).
/// Mirrors core/gurbanilens/corpus.py:Corpus
public final class Corpus {
    public let dbPath: URL
    private var db: OpaquePointer?
    private let sggsSourceId: Int
    private let englishLanguageId: Int

    public enum CorpusError: Error {
        case dbOpenFailed(String)
        case sggsSourceNotFound
        case englishLanguageNotFound
    }

    public init(dbPath: URL) throws {
        self.dbPath = dbPath

        var handle: OpaquePointer?
        let flags = SQLITE_OPEN_READONLY | SQLITE_OPEN_URI
        let result = sqlite3_open_v2(dbPath.path, &handle, flags, nil)
        guard result == SQLITE_OK, let db = handle else {
            let msg = handle.flatMap { String(cString: sqlite3_errmsg($0)) } ?? "unknown"
            sqlite3_close(handle)
            throw CorpusError.dbOpenFailed(msg)
        }
        self.db = db

        // Resolve SGGS source id and English language id
        guard let id = Corpus.resolveId(db: db, table: "sources",
                                       where_: "name_english LIKE ? || '%'",
                                       value: "Sri Guru Granth Sahib") else {
            sqlite3_close(db)
            throw CorpusError.sggsSourceNotFound
        }
        self.sggsSourceId = id

        guard let lang = Corpus.resolveId(db: db, table: "languages",
                                         where_: "name_english = ?",
                                         value: "English") else {
            sqlite3_close(db)
            throw CorpusError.englishLanguageNotFound
        }
        self.englishLanguageId = lang
    }

    deinit {
        if let db = db { sqlite3_close(db) }
    }

    private static func resolveId(db: OpaquePointer, table: String,
                                  where_: String, value: String) -> Int? {
        let sql = "SELECT id FROM \(table) WHERE \(where_)"
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return nil }
        defer { sqlite3_finalize(stmt) }

        let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)
        sqlite3_bind_text(stmt, 1, value, -1, SQLITE_TRANSIENT)
        guard sqlite3_step(stmt) == SQLITE_ROW else { return nil }
        return Int(sqlite3_column_int64(stmt, 0))
    }

    // ---------------------------------------------------------------------
    // Iteration: all SGGS lines in canonical order
    // ---------------------------------------------------------------------

    /// Iterate every SGGS line that has English transliteration. Used by
    /// the matcher to build its in-memory index.
    public func allLines() throws -> [Line] {
        let sql = """
        SELECT l.id, l.shabad_id, l.source_page, l.source_line, l.gurmukhi,
               l.first_letters, l.order_id,
               lt.name_english AS line_type,
               t.transliteration AS transliteration_en
        FROM lines l
        JOIN shabads s ON l.shabad_id = s.id
        LEFT JOIN line_types lt ON l.type_id = lt.id
        LEFT JOIN transliterations t
            ON t.line_id = l.id AND t.language_id = ?
        WHERE s.source_id = ?
        ORDER BY l.order_id
        """
        return try query(sql) { stmt in
            sqlite3_bind_int64(stmt, 1, Int64(englishLanguageId))
            sqlite3_bind_int64(stmt, 2, Int64(sggsSourceId))
        }
    }

    public func lookup(ang: Int, pangti: Int) throws -> [Line] {
        let sql = """
        SELECT l.id, l.shabad_id, l.source_page, l.source_line, l.gurmukhi,
               l.first_letters, l.order_id,
               lt.name_english AS line_type,
               t.transliteration AS transliteration_en
        FROM lines l
        JOIN shabads s ON l.shabad_id = s.id
        LEFT JOIN line_types lt ON l.type_id = lt.id
        LEFT JOIN transliterations t
            ON t.line_id = l.id AND t.language_id = ?
        WHERE s.source_id = ? AND l.source_page = ? AND l.source_line = ?
        ORDER BY l.order_id
        """
        return try query(sql) { stmt in
            sqlite3_bind_int64(stmt, 1, Int64(englishLanguageId))
            sqlite3_bind_int64(stmt, 2, Int64(sggsSourceId))
            sqlite3_bind_int64(stmt, 3, Int64(ang))
            sqlite3_bind_int64(stmt, 4, Int64(pangti))
        }
    }

    /// All lines in a single Shabad, ordered. Used by the v1 ShabadScreen
    /// after a voice-search match.
    public func shabadLines(shabadId: String) throws -> [Line] {
        let sql = """
        SELECT l.id, l.shabad_id, l.source_page, l.source_line, l.gurmukhi,
               l.first_letters, l.order_id,
               lt.name_english AS line_type,
               t.transliteration AS transliteration_en
        FROM lines l
        LEFT JOIN line_types lt ON l.type_id = lt.id
        LEFT JOIN transliterations t
            ON t.line_id = l.id AND t.language_id = ?
        WHERE l.shabad_id = ?
        ORDER BY l.order_id
        """
        guard let db = db else { return [] }
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return [] }
        defer { sqlite3_finalize(stmt) }
        let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)
        sqlite3_bind_int64(stmt, 1, Int64(englishLanguageId))
        sqlite3_bind_text(stmt, 2, shabadId, -1, SQLITE_TRANSIENT)
        var lines: [Line] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            lines.append(rowToLine(stmt))
        }
        return lines
    }

    // ---------------------------------------------------------------------
    // Internal helpers
    // ---------------------------------------------------------------------

    private func query(_ sql: String,
                       bind: (OpaquePointer?) -> Void) throws -> [Line] {
        guard let db = db else { return [] }
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            return []
        }
        defer { sqlite3_finalize(stmt) }
        bind(stmt)

        var lines: [Line] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            lines.append(rowToLine(stmt))
        }
        return lines
    }

    private func rowToLine(_ stmt: OpaquePointer?) -> Line {
        // Column order matches the SQL above; no _unicode columns yet — the
        // Phase 2A build pipeline adds those in the app database, not in
        // the raw shabados/database used for port parity.
        let id = stringColumn(stmt, 0) ?? ""
        let shabadId = stringColumn(stmt, 1) ?? ""
        let ang = Int(sqlite3_column_int64(stmt, 2))
        let pangtiRaw = sqlite3_column_type(stmt, 3) == SQLITE_NULL
            ? nil
            : Optional(Int(sqlite3_column_int64(stmt, 3)))
        let gurmukhi = stringColumn(stmt, 4) ?? ""
        let firstLetters = stringColumn(stmt, 5)
        let orderId = Int(sqlite3_column_int64(stmt, 6))
        let lineType = stringColumn(stmt, 7)
        let transliterationEn = stringColumn(stmt, 8)
        return Line(
            id: id,
            shabadId: shabadId,
            ang: ang,
            pangti: pangtiRaw,
            lineType: lineType,
            gurmukhi: gurmukhi,
            gurmukhiUnicode: nil,
            transliterationEn: transliterationEn,
            firstLetters: firstLetters,
            orderId: orderId
        )
    }

    private func stringColumn(_ stmt: OpaquePointer?, _ index: Int32) -> String? {
        if sqlite3_column_type(stmt, index) == SQLITE_NULL { return nil }
        guard let ptr = sqlite3_column_text(stmt, index) else { return nil }
        return String(cString: ptr)
    }
}
